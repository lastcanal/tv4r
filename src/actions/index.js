import { extractPosts, filterPosts } from '../helpers'
import reddit from '../helpers/reddit'

import {
  SELECT_SUBREDDIT, INVALIDATE_SUBREDDIT,
  REQUEST_POSTS, RECEIVE_POSTS, RECEIVE_POSTS_ERROR,
  SELECT_POST, NEXT_POST, PREVIOUS_POST,
  MEDIA_FALLBACK
} from '../constants'

export const selectSubreddit = subreddit => ({
  type: SELECT_SUBREDDIT,
  subreddit
})

export const invalidateSubreddit = subreddit => ({
  type: INVALIDATE_SUBREDDIT,
  subreddit
})

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit
})

export const receivePosts = (subreddit, json) => ({
  type: RECEIVE_POSTS,
  subreddit,
  posts: filterPosts(extractPosts(json)),
  error: null,
  receivedAt: Date.now()
})

export const receivePostsError = (subreddit, error) => ({
  type: RECEIVE_POSTS_ERROR,
  subreddit,
  error
})


export const selectPost = (post, index) => ({
  type: SELECT_POST,
  post,
  index
})

export const nextPost = (posts) => ({
  type: NEXT_POST,
  posts
})

export const previousPost = (posts) => ({
  type: PREVIOUS_POST,
  posts
})

export const mediaFallback = () => ({
  type: MEDIA_FALLBACK
})

const fetchPosts = subreddit => dispatch => {
  dispatch(requestPosts(subreddit))
  return reddit.fetchPosts(subreddit)
    .then(response => response.json())
    .then(response => dispatch(receivePosts(subreddit, response)))
    .catch(error => dispatch(receivePostsError(subreddit, error)))
}

const shouldFetchPosts = (state, subreddit) => {
  const posts = state.postsBySubreddit[subreddit]
  if (!posts) {
    return true
  }
  if (posts.isFetching) {
    return false
  }
  return posts.didInvalidate
}

export const fetchPostsIfNeeded = subreddit => (dispatch, getState) => {
  if (shouldFetchPosts(getState(), subreddit)) {
    return dispatch(fetchPosts(subreddit))
  }
}

