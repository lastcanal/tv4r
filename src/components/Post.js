import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { isVideo } from '../helpers'
import ReactHtmlParser from 'react-html-parser';
import ReactPlayer from 'react-player'

import { nextPost, mediaFallback } from '../actions'

const styles = theme => ({
  root: {
    margin: 0,
    padding: 0,
  },
  playerWrapper: {},
  reactPlayer: {
    position: 'absolute',
    top: 0,
    left: 0,
  },
  spacer: {
    height: '100vh'
  }
});

class Post extends Component {

  mediaEmbedContent() {
    return new DOMParser()
      .parseFromString(this.props.post.media_embed.content, "text/html")
      .documentElement
      .textContent;
  }

  renderMediaEmbed() {
    const { post, classes } = this.props
    if (post && isVideo(post)) {
      const transform = (node, index) => {
        if(node.type === 'tag' && node.name === 'iframe') {
          node.attribs.height = "100%"
          node.attribs.width = "100%"
          node.attribs.class = classes.reactPlayer
        }
      }
      return <div class={classes.playerWrapper}>
        {ReactHtmlParser(this.mediaEmbedContent(), {transform})}
      </div>;
    } else {
      return <div></div>
    }
  }

  renderMediaPlayer() {
    const { post, classes } = this.props
    return <div class={classes.playerWrapper}>
      <ReactPlayer
        ref={node => this.player = node}
        playing
        url={post.url}
        className={classes.reactPlayer}
        width='100%'
        height='100%'
        onEnded={this.onMediaEnded.bind(this)}
        onError={this.onMediaError.bind(this)}
        controls={true} />
    </div>
  }

  renderMedia() {
    const { media_fallback } = this.props
    if (media_fallback) {
      return this.renderMediaEmbed()
    } else {
      return this.renderMediaPlayer()
    }
  }

  onMediaEnded() {
    const { dispatch, posts } = this.props
    dispatch(nextPost(posts))
  }

  onMediaError(error) {
    const { dispatch } = this.props
    dispatch(mediaFallback())
  }

  renderLoading() {
    return <h2>Loading...</h2>
  }

  renderEmpty() {
    return <h2>No TV Found.</h2>
  }

  render() {
    const { classes, posts, isFetching } = this.props
    if (isFetching) {
      return this.renderLoading()
    } else if (posts.length === 0) {
      return this.renderEmpty()
    } else {
      return <div style={{ opacity: isFetching ? 0.5 : 1 }}>
        <Container maxWidth={false} className={classes.root}>
          <div>{this.renderMedia()}</div>
        </Container>
        <div class={classes.spacer}></div>
      </div>
    }
  }
}

Post.propTypes = {
  post: PropTypes.object.isRequired
}

const mapStateToProps = state => {
  const {
    dispatch,
    selectedSubreddit,
    postsBySubreddit,
    selectedPost } = state

  const {
    isFetching,
    items: posts,
  } = postsBySubreddit[selectedSubreddit] || {
    items: []
  }

  return {
    dispatch,
    posts,
    isFetching,
    post: selectedPost.post,
    media_fallback: selectedPost.media_fallback
  }
}

export default connect(mapStateToProps)(withStyles(styles)(Post))



