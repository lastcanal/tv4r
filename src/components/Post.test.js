import { Provider } from 'react-redux'

import Post from './Post'
import * as actions from '../actions'

describe('Post', () => {

  it('should render empty Post', () => {
    const wrapper = mount(
      <Provider store={makeStore()}>
        <Post />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render Post', () => {
    const post = {
      id: 1,
      url: "https://example.com/video"
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [ post ],
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render media fallback', () => {
    const post = {
      id: 1,
      url: "https://example.com/video",
      media: {
        oembed: {
          type: 'video'
        }
      },
      media_embed: {
        content: "&lt;iframe width=\"600\" height=\"338\" src=\"https://example.com/video\" frameborder=\"0\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen&gt;&lt;/iframe&gt;"
      }
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [ post ],
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post, media_fallback: true }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should not render media fallback for non-videos', () => {
    const post = {
      id: 1,
      url: "https://example.com/video",
      media: {
        oembed: {
          type: 'image'
        }
      },
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [ post ],
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post, media_fallback: true }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render Post is loading', () => {
    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [  ],
          isFetching: true,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: {  }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render next post', () => {
    const post1 = {
      id: 1,
      url: "https://example.com/video",
    }

    const post2 = {
      id: 2,
      url: "https://example.com/video2",
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [ post1, post2 ],
          isFetching: false,
          didInvalidate: false
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post: post1 }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    store.dispatch(actions.nextPost())
    expect(wrapper).toMatchSnapshot()
  })


  it('should show loading when isFetching', () => {
    const post = {
      id: 1,
      url: "https://example.com/video",
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [ post ],
          isFetching: true,
          didInvalidate: true
        }
      },
      selectedSubreddit: 'foo',
      selectedPost: { post }
    })


    const wrapper = mount(
      <Provider store={store}>
        <Post />
      </Provider>
    )

    store.dispatch(actions.nextPost())
    expect(wrapper).toMatchSnapshot()
  })

})
