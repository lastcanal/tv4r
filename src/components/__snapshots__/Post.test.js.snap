// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should not render media fallback for non-videos 1`] = `ReactWrapper {}`;

exports[`Post should render Post 1`] = `ReactWrapper {}`;

exports[`Post should render Post is loading 1`] = `ReactWrapper {}`;

exports[`Post should render empty Post 1`] = `ReactWrapper {}`;

exports[`Post should render media fallback 1`] = `ReactWrapper {}`;

exports[`Post should render next post 1`] = `ReactWrapper {}`;

exports[`Post should show loading when isFetching 1`] = `ReactWrapper {}`;
